lexer/stateful/pattern.go

```diff
@@ -0,0 +1,218 @@
+package stateful
+
+import (
+	"fmt"
+	"regexp/syntax"
+	"unicode"
+)
+
+///////////////////////////////////////////
+// These functions analyze a Regexp AST to determine for a given pattern ;
+//   - what possible first runes it would match on
+//   - what possible match size can be expected (possibly zero, exactly one, one or more)
+// The possible match size is for now not utilized, but could potentially be used to not
+// even launch a regexp match if the rune is found is in the input.
+///////////////////////////////////////////
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// If a character class has a range bigger than this, then give up on it and mark it as always to test,
+// since otherwise it would take too much memory.
+const (
+	charclassSizeLimit = 4096
+	mayMatchZero       = 0
```

Comment:
BTW for future reference this incrementing constants is solved by iota:

```go
const (
  mayMatchZero = iota
  mayMatchOneOrMore
  matchesExactlyOne
)
```

(yes, the syntax is bizarre)